text/0000-template-comments.md

```diff
@@ -0,0 +1,86 @@
+---
+title: HTML comments for LWC templates
+status: DRAFTED
+created_at: 2021-03-31
+updated_at: 2021-03-31
+pr: https://github.com/salesforce/lwc/pull/2263
+---
+
+# HTML comments for LWC templates
+
+## Summary
+
+In most cases, HTML comments in templates only add information for other developers. There are some use cases where HTML comments convey some runtime semantic (e.g., MSO comments). As of today, the LWC template compiler strips HTML comments. This RFC proposes a way to preserve HTML comments in LWC templates.
+
+### Basic example
+
+```html
+<template preserve-comments>
+  <!-- Greeter container -->
+  <div>
+	Hello <b>{name}</b>!
+  </div>
+  <!-- eof: Greeter container -->
+</template>
+```
+```html
+<x-example>
+  # shadow-root
+  |   <!-- Greeter container -->
+  |   <div>
+  |     Hello <b>world</b>
+  |   </div>
+  |   <!-- eof: Greeter container -->
+</x-example>
+```
+
+## Motivation
+
+Microsoft Outlook's desktop email client leverage MSO (Microsoft Office) to render their HTML content. Because of this, the support for common HTML and CSS properties is limited, eg: `<div>`s are not supported.
+
+Email developers rely on [MSO conditionals](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/compatibility/ms537512(v=vs.85)?redirectedfrom=MSDN)  to improve the UX of emails opened on Microsoft Outlook.
+
+This RFC introduces a new opt-in mechanism at compile time to preserve comments on the rendered component.
+
+## Proposal
+
+### Compile-time behavior
+
+LWC compilation strips HTML comments because the browsers do not use them at runtime. The bundle's size is smaller by stripping the HTML comments, and the rehydration process does not need to take into account comments nodes.
+
+To preserve the performance of the existing components, we propose to introduce one of these:
+
+1. A new attribute (`preserve-comments`) to the root template tag in the component template.
+
+2. Add a compile option (`preserveHTMLComments`) to the template compiler.
+
+Both options will make the compiler invoke a runtime API adding the HTML comments in the component template to the output code that runs in the browser/server. Modifications in each environment's renderers will ensure that the comment node/text is present on the resulting HTML.
+
+### Runtime behavior
+
+A new API "`[co]`mment" will be added to the engine/core. The `co(string commentText): VNode` will receive the comment text and return a Comment vnode.
+
+During rendering cycles, the LWC engine diffing algorithm will show/hide the comment based on the directive containing it (if:true/false, for:each, etc.)
+
+#### Invariants
+
+* HTML comments will be supported based on the opt-in mechanism (TBD, either option 1 or 2).
+* HTML comments will not support bindings (`{foo}`).
```

Comment:
In the example above the binding is no on the HTML comment itself. There are 2 HTML comment nodes surrounding a `<i>` element with a `style` attribute binding. Here is as an example of an HTML comment binding:

```html
<template>
  <!-- With dynamic content: {content} -->
</template>
```

The `{content}` will not be interpreted as an expression but as a static text content.

Reply:
Am not sure inner html binding would work in this case since we are not injecting html content here. Rather the binding is to set the attribute value of the <i> tag in this example. The actual value thats being rendered today for the binding in the above example is 
`<i style="mso-font-width:-100%;letter-spacing:20px;mso-text-raise:-10px;..>`

Reply:
Here is an example of how declarative inner HTML binding could be used to injected dynamic HTML comments:

```html
<template>
  <div lwc:inner-html={msoComment}></div>
</template>
```

```js
import { LightningElement } from 'lwc';

export default class extends LightningElement {
  get msoComment() {
    const iLeftStyle = '... some computed content ...';
    return `<!--[if mso]><i style=${iLeftStyle}>&nbsp;</i><![endif]-->`;
  }
}
```

This main constrain with declarative inner HTML binding is that it is bound to an element and can't be injected directly inside the template body. As far as I understand how MSO comment works, it should be fine to inject it in a `<div>`.

Reply:
Got it thanks for the example. I see that the inner html RFC is also in draft state. Would it be implemented by then ?